### DISCUSSION TITLE
Ways to test individual application

### DISCUSSION THREAD

**Discussion by ivanperez-keera** [nasa/cFS#507] [Q&A]
We are writing individual monitoring apps for cFS, and we'd like to test them.

The way we are doing it now is by adding some applications that can replay a complete run, but that means we'd be running many applications in the system even if we just want to test one. It also means that I have to inspect a log/terminal manually to see that my application is doing the right thing, but the output contains also data from other applications running simultaneously.

The kinds of properties I envision testing are of the following shape:
- If a message of X kind with data Y comes in the bus, the application's status switches to Z (provided that it can be inspected).
- If a message of X kind with data Y comes in the bus, the application reports Z to the log in W time.
- If a message of X1 kind with data Y1 comes in the bus, and then a message of kind X2 with data Y2, the application's reports Z to the log in W time.
- If a message of kind X1 with data Y1 comes in the bus, and then a message of kind X2 with data Y2, the application's stops in W time.

and so on.

What is the best way to test an individual app?

I see that there is [this discussion](https://github.com/nasa/cFS/discussions/308), but it never reached a conclusion and it doesn't answer this question, as far as I can see.

**Comment by skliper** [comment]
We use UtAssert for coverage (stubbed) testing and test application based functional/api testing.  See the sample-app unit test directory as an example of coverage testing (https://github.com/nasa/sample_app/tree/main/unit-test), or any of the coverage testing in cFE, OSAL, or many of the open source GSFC apps.  The test application for functional/api testing of cFE is here: https://github.com/nasa/cFE/tree/main/modules/cfe_testcase.  We've recently started using CTF (https://github.com/nasa/ctf) along with CCDD (https://github.com/nasa/ccdd) for what you describe above.  We do plan to eventually release those tests, but we haven't generalized them for open source yet (they are still project specific).

**Reply by ivanperez-keera** [reply]
@skliper Any updates? Any chance I'd be able to see those tests? You can also contact me via NASA email (I'm in the directory as Ivan Perez Dominguez.) Thanks!

**Reply by skliper** [reply]
@ivanperez-keera - I no longer support cFS directly, but those tests have come a long ways.  I recommend emailing [cfs-program@lists.nasa.gov](mailto:cfs-program@lists.nasa.gov), to request access and/or update on the status of transitioning those to open source.
